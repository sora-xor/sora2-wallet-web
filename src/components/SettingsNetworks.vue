<template>
  <wallet-base :title="title" show-back @back="handleBack">
    <div class="wallet-settings-networks">
      <template v-if="step === 1">
        <span class="wallet-settings-networks_desc">{{ t('settings.networks.desc') }}</span>
        <div
          v-for="(net, index) in availableNetworks"
          :key="index"
          @click="openNetwork(net)"
        >
          <div class="wallet-settings-networks-item s-flex">
            <div class="wallet-settings-networks-item-text s-flex">
              <span class="wallet-settings-networks-item-text_main">{{ net.name }}</span>
            </div>
            <s-icon
              class="wallet-settings-networks-item_icon"
              name="chevron-right"
              :size="16"
            />
          </div>
          <s-divider v-if="index !== availableNetworks.length - 1" class="wallet-settings-networks-item_divider" />
        </div>
        <div class="wallet-settings-networks_action s-flex">
          <s-button
            type="secondary"
            size="small"
            @click="createNetwork"
          >
            {{ t('settings.networks.create') }}
          </s-button>
        </div>
      </template>
      <template v-else>
        <s-form
          :model="network"
          :disabled="!network.editable"
        >
          <s-form-item prop="name">
            <s-input v-model="network.name" :placeholder="t('settings.networks.form.name')" />
          </s-form-item>
          <s-form-item prop="address">
            <s-input v-model="network.address" :placeholder="t('settings.networks.form.address')" />
          </s-form-item>
          <s-form-item prop="explorer">
            <s-input v-model="network.explorer" :placeholder="t('settings.networks.form.explorer')" />
          </s-form-item>
        </s-form>
        <div class="wallet-settings-networks_action-group s-flex">
          <s-button
            type="secondary"
            size="small"
            @click="onCancel"
          >
            {{ t('cancelText')}}
          </s-button>
          <s-button
            :disabled="!network.editable"
            type="primary"
            size="small"
            @click="onSave"
          >
            {{ t('saveText') }}
          </s-button>
        </div>
      </template>
    </div>
  </wallet-base>
</template>

<script lang="ts">
import { Component, Mixins, Prop } from 'vue-property-decorator'
import { Action, Getter, State } from 'vuex-class'
import last from 'lodash/last'

import TranslationMixin from './mixins/TranslationMixin'
import WalletBase from './WalletBase.vue'
import { Network, RouteNames } from '../consts'

@Component({
  components: {
    WalletBase
  }
})
export default class SettingsNetworks extends Mixins(TranslationMixin) {
  @Getter availableNetworks

  @Action navigate
  @Action addNetwork

  step = 1
  network: Network = {
    id: 0,
    name: '',
    address: '',
    explorer: '',
    editable: false
  }

  get title (): string {
    if (this.step === 2) return this.network.name || this.t('settings.networks.customNetwork')
    return this.t('settings.menu.Networks.title')
  }

  openNetwork (net: Network): void {
    this.step = 2
    this.network = { ...net }
  }

  createNetwork (): void {
    const lastNetwork: Network | undefined = last(this.availableNetworks)
    const id = lastNetwork ? (lastNetwork.id + 1) : 1
    this.step = 2
    this.network = {
      id,
      name: '',
      address: '',
      explorer: '',
      editable: true
    }
  }

  onCancel (): void {
    this.step = 1
  }

  onSave (): void {
    this.addNetwork({ network: this.network })
    this.step = 1
  }

  handleBack (): void {
    if (this.step === 1) {
      this.navigate({ name: RouteNames.WalletSettings })
    } else {
      this.onCancel()
    }
  }
}
</script>

<style scoped lang="scss">
.wallet-settings-networks {
  &_desc {
    color: var(--s-color-base-content-tertiary);
    font-size: $font-size_small;
    padding: $basic-spacing_small 0;
  }
  &_action {
    justify-content: center;
    > * {
      width: 100%;
    }
  }
  &_action-group {
    & > button {
      width: 100%;
    }
  }
  &-item {
    justify-content: space-between;
    padding: $basic-spacing_small 0;
    &-text {
      flex-direction: column;
      &_main {
        text-decoration: unset;
        color: unset;
        font-size: $font-size_normal;
        line-height: 1.8;
      }
    }
    &_icon {
      margin: auto 0;
    }
    &_divider {
      margin: unset;
    }
    &:hover {
      cursor: pointer;
      .wallet-settings-networks-item-text_main,
      .wallet-settings-networks-item_icon {
        color: var(--s-color-button-tertiary-color);
      }
    }
  }
}
</style>
