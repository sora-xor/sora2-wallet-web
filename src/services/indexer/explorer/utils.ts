import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
