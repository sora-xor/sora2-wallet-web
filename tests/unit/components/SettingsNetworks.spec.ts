import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils'

import SettingsNetworks from '@/components/SettingsNetworks.vue'
import { TranslationMock, SoramitsuElementsImport } from '../../utils'

const localVue = createLocalVue()
localVue.use(Vuex)
const store = new Vuex.Store({
  modules: {
    Router: {
      actions: {
        navigate: jest.fn()
      }
    },
    Settings: {
      getters: {
        availableNetworks: () => ([
          {
            id: 1,
            name: 'Main Ethereum Network',
            address: 'https://api.infura.io/v1/jsonrpc/mainnet',
            explorer: 'https://etherscan.io',
            editable: false
          }
        ])
      },
      actions: {
        addNetwork: jest.fn()
      }
    }
  } as any
})

describe('SettingsNetworks.vue', () => {
  beforeEach(() => {
    SoramitsuElementsImport(localVue)
    TranslationMock(SettingsNetworks)
  })

  it('should renders correctly', () => {
    const wrapper = shallowMount(SettingsNetworks, { localVue, store })
    expect(wrapper.element).toMatchSnapshot()
  })
})
