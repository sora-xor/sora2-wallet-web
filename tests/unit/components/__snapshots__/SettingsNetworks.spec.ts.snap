// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsNetworks.vue should renders correctly 1`] = `
<wallet-base-stub
  showback="true"
  title=""
>
  <div
    class="wallet-settings-networks"
  >
    <span
      class="wallet-settings-networks_desc"
    >
      
    </span>
     
    <div>
      <div
        class="wallet-settings-networks-item s-flex"
      >
        <div
          class="wallet-settings-networks-item-text s-flex"
        >
          <span
            class="wallet-settings-networks-item-text_main"
          >
            Main Ethereum Network
          </span>
        </div>
         
        <s-icon-stub
          class="wallet-settings-networks-item_icon"
          name="chevron-right"
          size="12px"
        />
      </div>
       
      <s-divider-stub
        class="wallet-settings-networks-item_divider"
        contentposition="center"
        direction="horizontal"
        type="secondary"
      />
    </div>
    <div>
      <div
        class="wallet-settings-networks-item s-flex"
      >
        <div
          class="wallet-settings-networks-item-text s-flex"
        >
          <span
            class="wallet-settings-networks-item-text_main"
          >
            Ropsten Network
          </span>
        </div>
         
        <s-icon-stub
          class="wallet-settings-networks-item_icon"
          name="chevron-right"
          size="12px"
        />
      </div>
       
      <!---->
    </div>
     
    <div
      class="wallet-settings-networks_action s-flex"
    >
      <s-button-stub
        icon=""
        nativetype="button"
        size="small"
        tabindex="0"
        tooltip=""
        tooltipplacement="top"
        type="secondary"
      >
        
          
        
      </s-button-stub>
    </div>
  </div>
</wallet-base-stub>
`;
